File: rd-core/src/commonMain/kotlin/com/jetbrains/rd/util/threading/SynchronousScheduler.kt
package com.jetbrains.rd.util.threading

import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.threadLocalWithInitial


object SynchronousScheduler : IScheduler {
    override fun flush() {

    }

    override fun queue(action: () -> Unit) {
        active.set(active.get() + 1)
        try {
            action()
        } finally {
            active.set(active.get() - 1)
        }
    }

    private val active =  threadLocalWithInitial { 0 }

    override val isActive: Boolean
        get() = active.get() > 0



}

File: rd-net/RdCore/Collections/Viewable/SynchronousScheduler.cs
using System;
using JetBrains.Lifetimes;

namespace JetBrains.Collections.Viewable
{
  public class SynchronousScheduler : IScheduler
  {
    public static SynchronousScheduler Instance = new SynchronousScheduler();

    [ThreadStatic] private static int ourActive;

    public void SetActive(Lifetime lifetime)
    {      
      lifetime.Bracket(() => { ourActive++; }, () => { ourActive--; });
    }

    private SynchronousScheduler(){}

    public void Queue(Action action)
    {
      try
      {
        ourActive ++;
        action();
      }
      finally
      {
        ourActive--;
      }
    }

    public bool IsActive { get { return ourActive > 0; } }
    public bool OutOfOrderExecution { get { return false; } }
  }
}